### DISCUSSION TITLE
ZEN-Garden as a Package

### DISCUSSION THREAD

**Discussion by jafluri** [ZEN-universe/ZEN-garden#183] [Installation & Setup] [answered]
As discussed in the last two meetings, we would like to transform the repo into a proper python package. I have implemented a basic version into the `development_JF` branch that includes the feedback from the last meeting. I recommend cloning the package into a temporary directoy if you want to try it out and play around before it is merged into the `development` branch. This is because the installation of the package will generate some files that are not in the `.gitignore` of the other branches, so when you checkout another branch after the package installation, it could become a bit messy. Obviously, once everything is merged, this won't be an issue anymore.

### Directory structure and setup

The biggest change is the structure of the directory. The directories `model`, `postprocess` and  `preprocess` are now in the package directory `zen_garden`. Additionally, there is a new file, `setup.py`, in the base directory that is used for the installation. To install the package into yout `zen-garden` environment, simply run 

```
pip install -e .
```

This will install the package such that all changes you peform in the code of the package will be immediately adopted at runtime without the need to install the package again. The `setup.py` file is also the place where the authors of the package, as well as, licences, keywords etc. are placed. Currently, it only contains the most basic information and a lot of placeholders that can be changed to the correct values later. 

### Changes to the package files

Inside the `zen_garden` directory, most files are basically the same. The main exceptions are the imports. Most imports are now relative to the package structure. Additionally, small adaptations in the path handling and other things have been done, but they should not affect how the package is used. There are additional files that allow one to run the package via `python -m zen_garden` (see below) and a couple of directories got an `__init__.py`. Otherwise, one of the biggest change is the `default_config.py`, where the default config was changed from loose dictionaries to a class called `Config` to avoid monkey patching. The usage of this new class is explained below.

### Changes in the usage

Previously, the repo was used by running the `compile.py` script. This script has been absorbed into the package.  The new usage is to run the package as a module via 

```
python -m zen_garden
```

This should be generally done inside the `data` folder where the `config.py` file is. However, the module accepts arguments. One can display the help message via `python -m zen_garden --help`. There are two supported arguments: `--config` and `--dataset`. The `--config` argument allows one to specify the path to the config file, if one does not run the package in the same directory and the `--dataset` argument allows one to overwrite the `config.analysis['dataset']` attribute of the config without explicitly changing the file. 

To run the repo as a package, one has to perform minor changes to the `config.py`, `system.py` and `scenarios.py` files. An example `config.py` looks like this:

```python
from zen_garden.model import Config
import os

# create a config
config = Config()

## Analysis - Default dictionary
analysis = config.analysis
## Solver - Default dictionary
solver = config.solver
## Scenarios - Default scenario dictionary
scenarios = config.scenarios

## Analysis - settings update compared to default values
analysis["dataset"]                                 = os.path.join(os.path.dirname(__file__), "test_1a")
analysis["objective"]                               = "TotalCost"                     # choose from "TotalCost", "TotalCarbonEmissions", "Risk"
analysis["useExistingCapacities"]                   = True                           # use greenfield or brownfield approach
## Solver - settings update compared to default values
#
solver["name"] = "glpk" # free solver
# solver["solverOptions"]["Method"]       = 2
# solver["solverOptions"]["NodeMethod"]   = 2
solver["solverOptions"]["BarHomogeneous"]   = 1
# solver["solverOptions"]["Presolve"]     = -1
solver["solverOptions"]["Threads"]      = 46
# solver["solverOptions"]["CrossoverBasis"]   = 0
# solver["solverOptions"]["Crossover"]    = 0
solver["solverOptions"]["ScaleFlag"]    = 2
solver["analyzeNumerics"]               = True
solver["immutableUnit"]                 = ["hour","km"]
```

The most important difference are:
- The import of the `Config` class at the top and the init of the config as `config = Config()`. **This line is not optional, the instance has to be named `config`, other names will lead to errors!**
- The `analysis["dataset"] ` should, whenever possible, be given as absolute path. Relative paths would work as well, but are not recommended.
- The imports of `system.py` and `scenarios.py` are now done by the package itself, based on the `analysis["dataset"] ` variable. There is no need to import them into the config.

An example of a `system.py` file is given here

```python
system = dict()
##############################

## System - settings update compared to default values
system['setConversionTechnologies']     = ["natural_gas_boiler"]
system['setStorageTechnologies']        = ["natural_gas_storage"]
system['setTransportTechnologies']      = ["natural_gas_pipeline"]

system['setNodes']                      = ["DE","CH"]
system["socialDiscountRate"]            = 0     # similar to discount factor, but for discounted utility model
system["knowledgeSpilloverRate"]        = 0.025
# time steps
system["referenceYear"]                 = 2022
system["unaggregatedTimeStepsPerYear"]  = 1
system["aggregatedTimeStepsPerYear"]    = 1
system["conductTimeSeriesAggregation"]  = False

system["optimizedYears"]                = 1
system["intervalBetweenYears"]          = 1
system["useRollingHorizon"]             = False
system["yearsInRollingHorizon"]         = 1
```
The most important difference is:
- There is no need import anything. The changes to can be written into a standard python dictionary. **The name `system` of this dictionary is not optional, other names will lead to errors!**

Finally, an example a `scenarios.py` file is given by

```python
scenarios                 = dict()
scenarios["2"]            = {"EnergySystem": ["carbonEmissionsLimit"]}
scenarios["3"]            = {"natural_gas": ["demandCarrier"]}
```

The most important difference is the same as for the `system.py` file:
- There is no need to import anything. The changes can be written into a standard python dictionary. **The name `scenarios` of this dictionary is not optional, other names will lead to errors!**

### Changes to the tests

All tests are now executed via the `tests/testcases/run_test.py` file. A test inside the file follows the following template:

```python
def test_1a(config):
    # run the test
    restore_default_state()
    compile(config=config, dataset_path=os.path.join(os.path.dirname(__file__), "test_1a"))
```

The most important aspects are:
- The name of any test has to start with "test_", otherwise it won't be executed by the pytest commands below.
- The input argument is provided by a pytest fixture, the config corresponds to the config inside `tests/testcases/config.py`
- The function `restore_default_state()` restores the initial state of the package and avoids monkey patching, it should be called at the beginning of any test.
- The path to the dataset used for the test should be given to the compile function as absolute path.
- Later, assertion test can simply be added to the individual functions. 

The execution of the tests is no platform independent and can be done by running:

```
coverage run --source="zen_garden" -m pytest -v tests/testcases/run_test.py
```

from the base directory of the repo. The coverage report is generated in the same way as before with

```
coverage report -m
```

**Comment by jafluri** [comment, marked as answer]
## PyCharm Setup

To easily execute the module with the PyCharm IDE one can use the configuration setup which can be found next to the run button:

![image](https://user-images.githubusercontent.com/15312314/193750411-ed79712e-6cfd-4b98-92c3-9551a622a4d0.png)

Click on "Edit Configurations..." to edit or add your configurations. There are three configurations that are eays to setup to run the module, the tests and get the coverage reports.

### Running the module

Add a new configuration by clicking on the "+" button on the top left corner of the window. Choose "Python" as a type. You give it any name you want. The imporant settings are:
- Change "Script Path" to "Module name" and set it to "zen_garden"
- Set the python interpreter to the Conda environment that was used to install the requirements and also has the package installed. **IMPORTANT:** This setup will only work for Conda environments that were also declared as such in PyCharm, if you set the path to the python executable yourself, you should create a new proper PyCharm interpretor.
- Set the "Working directory" to the path that contains the `config.py` file, this directory will also be used for the output.

In the end it should look something like this:

![image](https://user-images.githubusercontent.com/15312314/193752981-1275ca86-77ed-43cf-8712-14f0bfd899ba.png)

Make sure that the symbol inside the red circle matches the Conda env symbol of the image!

### Running the tests

To run the tests, add another "Python" type configuration. The important settings are:

- Change "Script Path" to "Module name" and set it to "coverage"
- Set the "Parameters" to: `run --source="zen_garden" -m pytest -v run_test.py`
- Set the python interpreter to the Conda environment that was used to install the requirements and also has the package installed. **IMPORTANT:** This setup will only work for Conda environments that were also declared as such in PyCharm, if you set the path to the python executable yourself, you should create a new proper PyCharm interpretor.
- Set the "Working directory" to the directory tests/testcases of the repo.

In the end it should look something like this:

<img width="642" alt="Screenshot 2024-05-06 at 13 38 48" src="https://github.com/ZEN-universe/ZEN-garden/assets/63449992/b3acc866-fa57-44d5-91d3-a994253c4635">

### Getting the coverage report

To run the test and also get the coverage report, we use the pipeline settings of the configuration. Add another "Python" configuration and use the following settings:

- Change "Script Path" to "Module name" and set it to "coverage"
- Set the "Parameters" to `report -m`
- Set the python interpreter to the Conda environment that was used to install the requirements and also has the package installed. **IMPORTANT:** This setup will only work for Conda environments that were also declared as such in PyCharm, if you set the path to the python executable yourself, you should create a new proper PyCharm interpretor.
- Set the "Working directory" to the base directory of the repo.
- Click the "+" button on the "Before launch" setting and choose "Run Another Configuration" and select the "Run Tests" configuration from above.

In the end it should something like this:

![image](https://user-images.githubusercontent.com/15312314/193752451-d57a45bb-a454-42ff-a6c9-df0b44c4a634.png)

**Reply by jafluri** [reply]
Since 63787c80711c559df09ed2bf98a3eea6dce28ead `results.py` contains the `Results` class and has a main program. This can be used to read in the results after the run and allows you run them in debug mode. The PyCharm config is shown in the image below:

![image](https://user-images.githubusercontent.com/15312314/207543811-e0b3b8dd-c8b8-4f70-b6a8-904bbe8309d0.png)

Make sure that:

- You have"Script Path" and not "Module name"  as above and set it to your path of `results.py`
- The "Working directory" is the ZEN-graden data directory

Optionally, you can add the "Run Module" configuration from above into the "Before launch" section. Then you will always generate the results that you read in first. Otherwise it will read in the results from the directory that is specified in you `config.py` file.

**Comment by johburger** [comment]
You also need to make sure that the `config.py` file does not contain the following lines anymore: 

```
## System - load system configurations
spec    = importlib.util.spec_from_file_location("module", os.path.join(f"{analysis['dataset']}", "system.py"))
module  = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)
system  = module.system

## overwrite default system and scenario dictionaries
if system["conductScenarioAnalysis"]:
    assert os.path.exists(f"{analysis['dataset']}/scenarios.py"), f"scenarios.py is missing from selected dataset '{analysis['dataset']}"
    spec        = importlib.util.spec_from_file_location("module", f"{analysis['dataset']}/scenarios.py")
    module      = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    scenarios   = module.scenarios
```

The `config.py` file should only include the lines from the above screenshot, nothign more.

**Comment by jafluri** [comment]
Since 564352efd1bbc986ca7c7833977ebb0600e9f932 the 

```
pip install -e .
```

step is included in the environment generation. So if you freshly install the environment, you don't need to run it. However, running it anyway will not hurt.

**Comment by vinzenz-muser** [comment]
In order to set up the run module possibility in VS code you can follow the same steps as described in the documentation in order to install Conda and the zen-garden package. Once it is installed, open the folder containing the project in VS Code and follow these steps:

- First, make sure that the right interpreter is selected: Press `ctrl + shift + p` to open the command palette (if you're on Windows or Linux), enter `Python: Select interpreter` and make sure that the interpreter of the correct conda environment is selected.
- Create a new file in the folder `./.vscode/` called `launch.json` with the following content:
  ```
  {
      "configurations": [
          {
              "name": "Python: ZEN-Garden",
              "type": "python",
              "cwd":"<path to folder with config.py>",
              "request": "launch",
              "module": "zen_garden",
              "console": "integratedTerminal"
          }
      ]
  }
  ```
  and make sure to adapt the `cwd`-field to the folder that contains your `config.py`.
- Save the file, select "Run" -> "Run Without Debugging" from the menu and your case should be running.
